### PR TITLE
e2e benchmarking

### PR DESCRIPTION
**Pull request by aciddelgado**
End-to-end Benchmarking figures of Phi-2 Model with Onnxruntime-GenAI

Example command used: 
python benchmark_e2e.py -i {model folder} -b 1 -l 128 -g 256 -r 100 -w 10 -k 5 -o {output csv file name} -ep {cpu or cuda}

*CUDA metrics run on RTX 4090 and CPU metrics run on 13th Gen Intel(R) Core(TM) i9-13900.
*Used ONNXRuntime 1.17.1 and "kv_shared_past_present":true
*Metrics averaged after 10 warmup runs and 100 repetitions. Prompt length 128. Token generation length 256. Top K 5.

INT4 CUDA
| Metric                           | Value       |
|----------------------------------|-------------|
| Batch Size                       | 1           |
| Prompt Length                    | 128         |
| Tokenization Throughput (tps)    | 3173.208879 |
| Tokenization Latency (ms)        | 0.31513841  |
| Prompt Processing Throughput (tps)| 9259.896566 |
| Prompt Processing Latency (ms)   | 0.107992567 |
| Token Generation Throughput (tps)| 219.8155497 |
| Token Generation Latency (ms)    | 4.549268699 |
| Sampling Throughput (tps)        | 40870.18761 |
| Sampling Latency (ms)            | 0.024467712 |
| Wall Clock Throughput (tps)      | 206.5942866 |
| Wall Clock Time (s)              | 1.239143658 |

FP16 CUDA
| Metric                           | Value       |
|----------------------------------|-------------|
| Batch Size                       | 1           |
| Prompt Length                    | 128         |
| Tokenization Throughput (tps)    | 4366.500114 |
| Tokenization Latency (ms)        | 0.229016369 |
| Prompt Processing Throughput (tps)| 10645.01097 |
| Prompt Processing Latency (ms)   | 0.09394072  |
| Token Generation Throughput (tps)| 105.2307414 |
| Token Generation Latency (ms)    | 9.502926488 |
| Sampling Throughput (tps)        | 35380.08808 |
| Sampling Latency (ms)            | 0.028264486 |
| Wall Clock Throughput (tps)      | 100.1268128 |
| Wall Clock Time (s)              | 2.556757703 |

INT4 CPU
| Metric                           | Value         |
|----------------------------------|---------------|
| Batch Size                       | 1             |
| Prompt Length                    | 128           |
| Tokenization Throughput (tps)    | 2838.653729   |
| Tokenization Latency (ms)        | 0.35227967    |
| Prompt Processing Throughput (tps)| 102.4218404  |
| Prompt Processing Latency (ms)   | 9.763542578   |
| Token Generation Throughput (tps)| 12.40760638  |
| Token Generation Latency (ms)    | 80.59572244   |
| Sampling Throughput (tps)        | 2637.265058   |
| Sampling Latency (ms)            | 0.379180696   |
| Wall Clock Throughput (tps)      | 12.39690874   |
| Wall Clock Time (s)              | 20.65030931   |

FP32 CPU
| Metric | Value |
| --- | --- |
| Batch Size | 1 |
| Prompt Length | 128 |
| Tokenization Throughput (tps) | 5279.162651 |
| Tokenization Latency (ms) | 0.18942398 |
| Prompt Processing Throughput (tps) | 104.3425442 |
| Prompt Processing Latency (ms) | 9.583818445 |
| Token Generation Throughput (tps) | 5.544113095 |
| Token Generation Latency (ms) | 180.3715009 |
| Sampling Throughput (tps) | 3025.604685 |
| Sampling Latency (ms) | 0.330512444 |
| Wall Clock Throughput (tps) | 5.555478121 |
| Wall Clock Time (s) | 46.08064229 |

### CODE REVIEW

**Comment by yufenglee** [comment]
could you please add a new folder for benchmark?

**Comment by natke** [comment]
Add in wall-clock metric

File: benchmark/python/benchmark_e2e.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            "Tokenization Latency (ms)",
+            "Prompt Processing Throughput (tps)",
+            "Prompt Processing Latency (ms)",
+            "Token Generation Throughput (tps)",
+            "Token Generation Latency (ms)",
+            "Sampling Throughput (tps)",
+            "Sampling Latency (ms)",
+        ],
+    )
+    df = df.transpose()  # This line swaps the rows and columns
```
Comment (kunal-vaishnavi) [inline benchmark/python/benchmark_e2e.py:44]: Why is the transpose needed?

File: benchmark/python/benchmark_e2e.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+#
+# 2) Run this script with the desired arguments. Run benchmark_e2e.py -h for help.
+
+import onnxruntime_genai as og
+import time
+import argparse
+
+# Use input model to generate prompt
+def generate_prompt(model, tokenizer, prompt_length) -> str:
+    prompt = "a"
```
Comment (kunal-vaishnavi) [inline benchmark/python/benchmark_e2e.py:16]: Should we use an actual small prompt (e.g. "The quick brown fox")?

File: benchmark/python/benchmark_e2e.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    parser.add_argument('-b', '--batch_size', type=int, default=1, help='Number of sequences to generate in parallel')
+    parser.add_argument('-l', '--prompt_length', type=int, default=16, help='Number of tokens for prompt')
+    parser.add_argument('-g', '--generation_length', type=int, default=256, help='Number of tokens to generate after prompt')
+    parser.add_argument('-r', '--repetitions', type=int, default=10, help='Number of times to repeat the benchmark')
+    parser.add_argument('-w', '--warmup', type=int, default=5, help='Number of warmup runs before benchmarking')
+    parser.add_argument('-k', '--top_k', type=int, default=50, help='Top k tokens to sample from')
+    parser.add_argument('-p', '--top_p', type=float, default=1.0, help='Top p probability to sample with')
+    parser.add_argument('-o', '--output', type=str, default='genai_e2e', help='Output CSV file name or path')
+    parser.add_argument('-d', '--device', type=str, choices=['cpu', 'cuda'], default='cpu', help='Device to use, default is CPU, use CUDA for GPU')
```
Comment (kunal-vaishnavi) [inline benchmark/python/benchmark_e2e.py:180]: Since this script is only benchmarking ONNX Runtime GenAI, should we change `-d/--device` to `-ep/--execution_provider`?

File: benchmark/python/benchmark_e2e.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+            "Sampling Latency (ms)",
+        ],
+    )
+    df = df.transpose()  # This line swaps the rows and columns
+    df.to_csv(filename, header=False)
+    print(f"Results saved in {filename}!")
+
+def main(args):
+    # Get user arguments
+    num_repititions = args.repetitions
```
Comment (kunal-vaishnavi) [inline benchmark/python/benchmark_e2e.py:50]:
```suggestion
    num_repetitions = args.repetitions
```

File: benchmark/python/benchmark_e2e.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    tokenizer = model.create_tokenizer()
+
+    # Generate prompt
+    prompt = [generate_prompt(model, tokenizer, prompt_length)] * batch_size
+    if args.verbose: print("Running warmup runs...")
+    for i in range(args.warmup):
+        if args.verbose: print(f"Running warmup repetition {i+1}...")
+        tokens = tokenizer.encode_batch(prompt)
+        params = og.GeneratorParams(model)
+        params.max_length = prompt_length + generation_length
```
Comment (kunal-vaishnavi) [inline benchmark/python/benchmark_e2e.py:67]: I think getting the `tokens` can be done outside of the `for` loop. Can the `params` object be outside the loop as well?

Comment (kunal-vaishnavi) [inline benchmark/python/benchmark_e2e.py:67]: I think getting the `tokens` can be done outside of the `for` loop.

File: benchmark/python/benchmark_e2e.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        tokenize_end_time = time.time()
+        tokenize_times.append(tokenize_end_time - tokenize_start_time)
+
+        # Measure prompt processing
+        prompt_start_time = time.time()
+        generator.compute_logits()
+        prompt_end_time = time.time()
+        prompt_times.append(prompt_end_time - prompt_start_time)
+
+        sampling_start_time = time.time()
```
Comment (kunal-vaishnavi) [inline benchmark/python/benchmark_e2e.py:101]: For some measurements such as sampling where the time difference is very small, should we use [`time.perf_counter()`](https://docs.python.org/3/library/time.html#time.perf_counter) instead of `time.time()`?

File: benchmark/python/benchmark_e2e.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        generator.generate_next_token_top_k_top_p(args.top_k, args.top_p, 1.0)
```
Comment (kunal-vaishnavi) [inline benchmark/python/benchmark_e2e.py:102]: Can we set `temperature = 1.0` and pass `temperature` to `generate_next_token_top_k_top_p` so it is clear what `1.0` represents?

File: benchmark/python/benchmark_e2e.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+        avg_per_token_prompt_latency_ms, 
+        avg_token_gen_thrpt, 
+        avg_token_gen_latency_ms, 
+        avg_sampling_thrpt, 
+        avg_sampling_latency_ms,
+    ]]
+
+    # Add metrics to CSV
+    if args.verbose: print("Adding results to CSV")
+    filename = args.output + ".csv"
```
Comment (kunal-vaishnavi) [inline benchmark/python/benchmark_e2e.py:166]:
```suggestion
    filename = args.output
```

The `help` message for `-o/--output` says that it is either a CSV filename or a path to a CSV filename, so I think users can add the ".csv" to what they pass to `-o/--output`.

File: benchmark/python/benchmark_e2e.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+            generator.compute_logits()
+            generator.generate_next_token_top_k_top_p(args.top_k, args.top_p, 1.0)
+        if args.print_model_output: print(tokenizer.decode(generator.get_sequence(0).get_array()))
+
+    tokenize_times = []
+    prompt_times = []
+    token_gen_times = []
+    sampling_times = []
+    if args.verbose: print(f"Running benchmark for batch size = {batch_size}, prompt length = {prompt_length}")
+    for i in range(num_repititions):
```
Comment (kunal-vaishnavi) [inline benchmark/python/benchmark_e2e.py:80]: Should we use `tqdm` so the user can see which iteration is running and how long the benchmarking will take?

File: benchmark/python/benchmark_e2e.py (hunk 1 above)
Comment (aciddelgado) [reply benchmark/python/benchmark_e2e.py:44]: Makes the table more readable on most screens

File: benchmark/python/benchmark_e2e.py (hunk 2 above)
Comment (aciddelgado) [reply benchmark/python/benchmark_e2e.py:16]: Might be a different amount of tokens depending on tokenizer but may not be a problem... Output from this single token prompt looks good usually though

File: benchmark/python/benchmark_e2e.py (hunk 6 above)
Comment (aciddelgado) [reply benchmark/python/benchmark_e2e.py:101]: should we use it everywhere or just in the very small ones like sampling and tokenization?

Comment (kunal-vaishnavi) [reply benchmark/python/benchmark_e2e.py:101]: We can use it everywhere.

**Comment by yufenglee** [comment]
Could you please add a simple readme?

File: benchmark/python/benchmark_e2e.py [hunk 10]
```diff
@@ -0,0 +1,201 @@
+# This is an end-to-end benchmarking script for any ONNX model.
+#
+# Prerequisites: 
+# 0) Install onnxruntime-genai and onnxruntime
+#
+# 1) Use builder.py to build the desired ONNX model
+#
+# 2) For accurate wall clock numbers, modify genai_config.json to include the following:
+#    "min_length": {prompt_length + generation_length}
```
Comment (yufenglee) [inline benchmark/python/benchmark_e2e.py:9]: does the user need to specify the min_length? I think we can specify it in the code

**Review by yufenglee — APPROVED** [review]

**Review by kunal-vaishnavi — APPROVED** [review]
